### DISCUSSION TITLE
Websocket not working in main loop

### DISCUSSION THREAD

**Discussion by roistoFIN** [emscripten-core/emscripten#22526] [Q&A]
Hi!

I have a simple game project using among other classes emscripten and websockets.

The code is basically as followed:

```
#include <emscripten/html5.h>
#include <webgl/webgl2.h>
...
#include <emscripten/websocket.h>

...
class Context
{
public:
	int* Ws;
public:
	Context(int* ws) {
		Ws = ws;
	}
};
Context* g_context;
...

EM_BOOL game_tick(double t, void* arg) {
	Context* context = static_cast<Context*>(arg);

        [game and draw logics here]

	EMSCRIPTEN_RESULT result;
	result = emscripten_websocket_send_utf8_text(*(context->Ws), "Just testing");
        printf("loop, *(context->Ws): %i\n", *(context->Ws));
}
...

EM_BOOL onopen(int eventType, const EmscriptenWebSocketOpenEvent* websocketEvent, void* userData) {
	puts("onconnect");
}
EM_BOOL onerror(int eventType, const EmscriptenWebSocketErrorEvent* websocketEvent, void* userData) {...}
EM_BOOL onclose(int eventType, const EmscriptenWebSocketCloseEvent* websocketEvent, void* userData) {...}
EM_BOOL onmessage(int eventType, const EmscriptenWebSocketMessageEvent* websocketEvent, void* userData) {
	puts("onmessage");
	if (websocketEvent->isText) {
		printf("message: %s\n", websocketEvent->data);
	}
	return EM_TRUE;
}
...

int main() {
        [init graphics stuff etc.]

	EmscriptenWebSocketCreateAttributes ws_attrs = {
		"http://localhost:8765",
		NULL,
		EM_TRUE
	};

	EMSCRIPTEN_WEBSOCKET_T ws = emscripten_websocket_new(&ws_attrs);

	emscripten_websocket_set_onopen_callback(ws, NULL, onopen);
	emscripten_websocket_set_onerror_callback(ws, NULL, onerror);
	emscripten_websocket_set_onclose_callback(ws, NULL, onclose);
	emscripten_websocket_set_onmessage_callback(ws, NULL, onmessage);

        printf("main ws: %i\n", ws);

	Context context(&ws);
	g_context = &context;

	emscripten_request_animation_frame_loop(&game_tick, &context);

       [image loading and other stuff]
}
```

I compile with following command:
`em++ programcpp -lGL -sMIN_WEBGL_VERSION=2 -sMINIMAL_RUNTIME=1 --js-library library_program.js --js-library library_websocket.js -o program.html`

The problem is that Websockets connection is not sending messages from main loop (result = emscripten_websocket_send_utf8_text(*(context->Ws), "Just testing");). The program just outputs:
```
main ws: 1
loop, *(context->Ws): 0
onopen
```
Which indicates that there is something wrong with Ws pointer. The connection should be fine, since I'm seeing `ClosedOK: received 1001 (going away); then sent 1001 (going away)` server side messages after browser close.

If add to main loop after `EMSCRIPTEN_WEBSOCKET_T ws = emscripten_websocket_new(&ws_attrs);` :
```
	EMSCRIPTEN_RESULT result;
	result = emscripten_websocket_send_utf8_text(ws, "Just testing\n");
```

Then I see that the direct call to ws is working since execution gives me a runtime error:
`Uncaught (in promise) InvalidStateError: Failed to execute 'send' on 'WebSocket': Still in CONNECTING state.`

So my question is that what am I doing wrong or missing here? I've double checked all the pointers, but again it is C...

BR,
-Risto

**Comment by sbc100** [comment]
I'm not sure if this is the issue but you are creating a `Context` object on the stack when you do `Context context(&ws);`.   This object will be destroyed when main returns so you should not assign it to the global `g_context` which lives longer than main.    

You should always be careful when taking the address of stack objects like this.   Most likely you want to instead do `g_context = new Context context(&ws);`.

**Reply by roistoFIN** [reply]
Thank you for the answer!

I've used this [code](https://pastebin.com/f4s8hsUS) as an "inspiration". The pong code compiles and websocket communication works (if there was a server) OK. It has exactly the same definitions, pointers and addresses of g_context and ws as mine code.

The only differences I can see is that pong uses _emscripten_set_main_loop_arg_ and my code uses _emscripten_request_animation_frame_loop_. My loop thus returns EM_TRUEs after each itreation and pong has infinite loop.

Could this be the reason?

Again, let me apologize that I'm rather unexperienced C++ coder!

**Reply by sbc100** [reply]
I think the difference here is that when the `simulate_infinite_loop` argument of `emscripten_set_main_loop_arg` is set to true the C++ descructor for the `Context context` local is not run.

As I said, taking the address of locals is tricky.  It should be easy for you confirm if this is the issue by using `new Context()`.. instead of using a local.

**Reply by roistoFIN** [reply]
I wrote the code again so that this time websockets init is called as a function from main loop (basically the same thing you suggested). It works now. Thank you!
